### PR TITLE
HARVESTER: Cloud provider enhancement

### PR DESCRIPTION
**Pull request by n313893254**
### Summary
- Add LB pages
- Add IP pool pages
- Remove the vip-pool setting
- Add VM instance labels, which can configure on VM and template page

### Occurred changes and/or fixed issues

- https://github.com/harvester/harvester/issues/1580
- https://github.com/harvester/harvester/issues/1458
- https://github.com/harvester/harvester/issues/1762
- https://github.com/harvester/harvester/issues/1834

### Screenshot/Video

image

image

image

image

### CODE REVIEW

**Comment by WuJun2016** [comment]
@n313893254   Please resolve the conflict.

File: pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    },
+
+    showRemove() {
+      return !this.isView;
+    },
+
+    namespaceOptions() {
+      const namespaces = this.$store.getters['harvester/all'](NAMESPACE);
+
+      const out = namespaces.filter((namespace) => {
```
Comment (WuJun2016) [inline pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue:52]: Do we need to filter out some system namespaces here ?

File: pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    >
+      <span v-if="isView">
+        {{ row.project }}
+      </span>
+      <Select
+        v-else
+        v-model="row.project"
+        :options="projectOptions"
+        @input="update"
+      />
```
Comment (WuJun2016) [inline pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue:182]: `InputOrDisplay`

File: pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    },
+  },
+};
+</script>
+
+<template>
+  <div
+    class="pool-row"
+    :class="{
+      'show-project': !isSingleVirtualCluster,
```
Comment (WuJun2016) [inline pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue:167]: image

Perhaps "Project" should not be displayed here?

File: pkg/harvester/edit/loadbalancer.harvesterhci.io.loadbalancer/HealthCheck.vue [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    },
+
+    disabled: {
+      type:    Boolean,
+      default: false,
+    },
+  },
+
+  data() {
+    return { healthCheckEnabled: !!this.value.port };
```
Comment (WuJun2016) [inline pkg/harvester/edit/loadbalancer.harvesterhci.io.loadbalancer/HealthCheck.vue:36]: Is it appropriate to use `this.value.port` to determine if healthCheckEnabled is enabled?

File: pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/Range.vue [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            <span v-if="isView">
+              {{ row.subnet }}
+            </span>
+            <input
+              v-else
+              v-model="row.subnet"
+              type="text"
+              @input="queueUpdate"
+            />
+          </div>
```
[the remark this answers is not in the document]
Comment (n313893254) [reply pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/Range.vue:157]: image

image

File: pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue (hunk 2 above)
Comment (n313893254) [reply pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue:182]: Ditto

File: pkg/harvester/edit/loadbalancer.harvesterhci.io.loadbalancer/HealthCheck.vue (hunk 4 above)
Comment (n313893254) [reply pkg/harvester/edit/loadbalancer.harvesterhci.io.loadbalancer/HealthCheck.vue:36]: Yes, port is required after enabled

File: pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue (hunk 3 above)
Comment (n313893254) [reply pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue:167]: Done

File: pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue (hunk 1 above)
Comment (n313893254) [reply pkg/harvester/edit/loadbalancer.harvesterhci.io.ippool/PriorityRow.vue:52]: Done

File: pkg/harvester/config/harvester.js [hunk 6]
```diff
@@ -725,4 +744,50 @@ export function init($plugin, store) {
     },
     exact: false,
   });
+
+  configureType(HCI.LB, { showState: false });
```
[the remark this answers is not in the document]
Comment (n313893254) [reply pkg/harvester/config/harvester.js]: Done

File: pkg/harvester/config/harvester.js [hunk 7]
```diff
@@ -725,4 +744,50 @@ export function init($plugin, store) {
...
+    },
+    {
+      name:     'ipam',
+      labelKey: 'harvester.loadBalancer.ipam.label',
+      value:    'ipamDisplay',
+    },
+    AGE
+  ]);
+
+  configureType(HCI.IP_POOL, { showState: false });
```
[the remark this answers is not in the document]
Comment (n313893254) [reply pkg/harvester/config/harvester.js]: Done

File: pkg/harvester/edit/kubevirt.io.virtualmachine/index.vue [hunk 8]
```diff
@@ -425,6 +428,14 @@ export default {
 
       return out;
     },
+
+    syncLabels() {
+      const labels = this.value.metadata.labels;
+
+      if (labels) {
+        this.value.spec.template.metadata.labels = labels;
+      }
```
[the remark this answers is not in the document]
Comment (n313893254) [reply pkg/harvester/edit/kubevirt.io.virtualmachine/index.vue]: Done

File: pkg/harvester/list/kubevirt.io.virtualmachine.vue [hunk 9]
```diff
@@ -11,6 +11,56 @@ import { HCI } from '../types';
...
+    value:     'id',
+    sort:      ['realAttachNodeName'],
+    formatter: 'HarvesterNodeName',
+    labelKey:  'tableHeaders.node'
+  },
+  {
+    ...AGE,
+    sort: 'metadata.creationTimestamp:desc',
+  }
+];
```
[the remark this answers is not in the document]
Comment (n313893254) [reply pkg/harvester/list/kubevirt.io.virtualmachine.vue]: Done

File: pkg/harvester/detail/loadbalancer.harvesterhci.io.ippool.vue [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+    await allHash(hash);
+  },
+
+  computed: {
+    networks() {
+      const inStore = this.$store.getters['currentProduct'].inStore;
+      const networks = this.$store.getters[`${ inStore }/all`](NETWORK_ATTACHMENT);
+
+      return networks.filter(n => n?.id === this.value?.spec?.network?.vlan);
```
[the remark this answers is not in the document]
Comment (n313893254) [reply pkg/harvester/detail/loadbalancer.harvesterhci.io.ippool.vue]: Done

File: pkg/harvester/detail/loadbalancer.harvesterhci.io.loadbalancer.vue [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+      class="bordered-table"
+      :weight="98"
+    >
+      <SortableTable
+        :rows="vms"
+        :headers="vmHeaders"
+        key-field="id"
+        :row-actions="false"
+        :table-actions="false"
+        :search="false"
```
[the remark this answers is not in the document]
Comment (n313893254) [reply pkg/harvester/detail/loadbalancer.harvesterhci.io.loadbalancer.vue]: Done

**Review by WuJun2016 — APPROVED** [review]
LGTM, thanks

**Review by WuJun2016 — APPROVED** [review]
LGTM

**Pull request merged**